File: 0841-keys-and-rooms/0841-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set()
        stack = [0]
        while stack:
            curr_room = stack.pop()
            if curr_room not in visited:
                visited.add(curr_room)
                for key in rooms[curr_room]:
                    stack.append(key)
        return len(visited) == len(rooms)

File: 0841-keys-and-rooms/0841-keys-and-rooms.c
bool canVisitAllRooms(int** rooms, int roomsSize, int* roomsColSize) {
    bool* visited = (bool*)calloc(roomsSize , sizeof(bool));
    int* stack = (int*)malloc(roomsSize * 2000 * sizeof(int));
    int top = -1;
    stack[++top] = 0;
    int visitedCount = 0;
    while(top >= 0){
        int curr = stack[top--];
        if(!visited[curr]){
            visited[curr] = true;
            visitedCount++;
            for(int i = 0; i < roomsColSize[curr]; i++){
                int key = rooms[curr][i];
                if(!visited[key]){
                    stack[++top] = key;
                }
            }
        }
    }
    free(visited);
    free(stack);
    return visitedCount == roomsSize;
}